pilot/pkg/secrets/kube/multicluster.go

```diff
@@ -0,0 +1,122 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"fmt"
+	"sync"
+
+	"istio.io/istio/pilot/pkg/secrets"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/secretcontroller"
+	"istio.io/pkg/log"
+)
+
+// Multicluster structure holds the remote kube Controllers and multicluster specific attributes.
+type Multicluster struct {
+	remoteKubeControllers map[string]*SecretsController
+	m                     sync.Mutex // protects remoteKubeControllers
+	secretController      *secretcontroller.Controller
+	localCluster          string
+}
+
+var _ secrets.Controller = &Multicluster{}
+
+func NewMulticluster(client kube.Client, localCluster, secretNamespace string) *Multicluster {
+	m := &Multicluster{
+		remoteKubeControllers: map[string]*SecretsController{},
+		localCluster:          localCluster,
+	}
+	// Add the local cluster
+	if err := m.AddMemberCluster(client, localCluster); err != nil {
+		return nil
+	}
+	sc := secretcontroller.StartSecretController(client,
+		m.AddMemberCluster,
+		m.UpdateMemberCluster,
+		m.DeleteMemberCluster,
+		secretNamespace)
+	m.secretController = sc
+	return m
+}
+
+func (m *Multicluster) AddMemberCluster(clients kube.Client, key string) error {
+	stopCh := make(chan struct{})
+	log.Infof("initializing Kubernetes credential reader for cluster %v", key)
+	sc := NewSecretsController(clients, key)
+	m.m.Lock()
+	m.remoteKubeControllers[key] = sc
+	m.m.Unlock()
+	clients.RunAndWait(stopCh)
+	return nil
+}
+
+func (m *Multicluster) UpdateMemberCluster(clients kube.Client, key string) error {
+	if err := m.DeleteMemberCluster(key); err != nil {
+		return err
+	}
+	return m.AddMemberCluster(clients, key)
+}
+
+func (m *Multicluster) DeleteMemberCluster(key string) error {
+	m.m.Lock()
+	delete(m.remoteKubeControllers, key)
+	m.m.Unlock()
+	return nil
+}
+
+func (m *Multicluster) GetKeyAndCert(name, namespace string) (key []byte, cert []byte) {
+	// Prefer local cluster
+	if lk, lc := m.remoteKubeControllers[m.localCluster].GetKeyAndCert(name, namespace); lk != nil && lc != nil {
```

Comment:
Seems like this requires unique names across clusters. Generate should be able to pass `proxy.Metadata.ClusterID` and we can fallback to searching, or only fallback to the "local" cluster (which is sort of the same as the "config" cluster)

Reply:
I have update this - we read from proxy cluster and fallback to config cluster. auth is always in proxy cluster